File: adk-go/src/application/tools/formatters.go
package tools

import "fmt"

func FormatInvoice(data map[string]any) map[string]any {
	requiredKeys :=[]string{"invoice", "seller", "buyer", "line_items"}
	for _, k := range requiredKeys {
		if _, ok := data[k]; !ok {
			return map[string]any{"error": fmt.Sprintf("Brakujące sekcje w danych wejściowych: %s", k)}
		}
	}

	var linesCalculated []map[string]any
	rawItems, ok := data["line_items"].([]interface{})
	if !ok {
		return map[string]any{"error": "line_items musi być poprawną listą."}
	}

	for _, itemRaw := range rawItems {
		item, ok := itemRaw.(map[string]any)
		if !ok {
			continue
		}

		qty, _ := item["quantity"].(float64)
		price, _ := item["unit_price_net"].(float64)
		vat, _ := item["vat_rate"].(string)

		calc := CalculateLine(qty, price, vat)
		if calc.Error != "" {
			name, _ := item["name"].(string)
			return map[string]any{"error": fmt.Sprintf("Błąd w obliczeniach pozycji '%s': %s", name, calc.Error)}
		}

		item["net_total"] = calc.NetTotal
		item["vat_amount"] = calc.VatAmount
		item["gross_total"] = calc.GrossTotal
		linesCalculated = append(linesCalculated, item)
	}

	totals := CalculateTotals(linesCalculated)

	return map[string]any{
		"invoice":    data["invoice"],
		"seller":     data["seller"],
		"buyer":      data["buyer"],
		"line_items": linesCalculated,
		"totals":     totals,
	}
}

File: adk-typescript/src/application/tools/formatters.ts
import { calculateLine, calculateTotals } from './calculators';

export function formatInvoice(data: any): any {
    if (!data || typeof data !== 'object') {
        return { error: "Dane wejściowe muszą być obiektem JSON." };
    }

    const requiredKeys = ['invoice', 'seller', 'buyer', 'line_items'];
    for (const key of requiredKeys) {
        if (!(key in data)) {
            return { error: `Brakujące sekcje w danych wejściowych: ${key}` };
        }
    }

    if (!Array.isArray(data.line_items)) {
        return { error: "line_items musi być poprawną listą." };
    }

    const linesCalculated: any[] = [];

    for (const item of data.line_items) {
        if (!item || typeof item !== 'object') continue;

        const qty = typeof item.quantity === 'number' ? item.quantity : parseFloat(item.quantity) || 0;
        const price = typeof item.unit_price_net === 'number' ? item.unit_price_net : parseFloat(item.unit_price_net) || 0;
        const vatRate = typeof item.vat_rate === 'string' ? item.vat_rate : String(item.vat_rate || "23%");

        const calc = calculateLine(qty, price, vatRate);

        const calculatedItem = { ...item };
        calculatedItem.net_total = calc.net_total;
        calculatedItem.vat_amount = calc.vat_amount;
        calculatedItem.gross_total = calc.gross_total;

        linesCalculated.push(calculatedItem);
    }

    const totals = calculateTotals(linesCalculated);

    return {
        invoice: data.invoice,
        seller: data.seller,
        buyer: data.buyer,
        line_items: linesCalculated,
        totals: {
            net: totals.total_net,
            vat: totals.total_vat,
            gross: totals.total_gross
        }
    };
}